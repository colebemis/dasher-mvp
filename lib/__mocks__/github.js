export const getGithubToken = jest.fn(() => 'TEST_GITHUB_TOKEN')
