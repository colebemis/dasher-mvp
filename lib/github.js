import getConfig from 'next/config'
import { stringify } from 'querystring'
import axios from 'axios'
import { join } from './utils'

const { serverRuntimeConfig, publicRuntimeConfig } = getConfig()

function graphqlRequest({ query, variables = {}, githubToken }) {
  return axios({
    url: 'https://api.github.com/graphql',
    method: 'POST',
    headers: { Authorization: `bearer ${githubToken}` },
    data: { query, variables },
  })
}

export function getGithubAuthUrl(from = '/') {
  const queryString = stringify({
    client_id: publicRuntimeConfig.githubClientId,
    redirect_uri: publicRuntimeConfig.redirectUri + '?' + stringify({ from }),
  })

  return `https://github.com/login/oauth/authorize?${queryString}`
}

export async function getGithubToken(githubCode) {
  const { data } = await axios({
    url: 'https://github.com/login/oauth/access_token',
    method: 'POST',
    headers: {
      Accept: 'application/json',
    },
    data: {
      client_id: publicRuntimeConfig.githubClientId,
      client_secret: serverRuntimeConfig.githubClientSecret,
      code: githubCode,
    },
  })

  if (data.error) {
    throw new Error(data.error)
  }

  return data.access_token
}

export function searchPullRequests({
  githubQuery,
  endCursor = null,
  batchSize = 10,
  githubToken,
}) {
  const query = `
    query($githubQuery: String!, $endCursor: String, $batchSize: Int) {
      search(
        query: $githubQuery,
        first: $batchSize,
        type: ISSUE,
        after: $endCursor
      ) {
        issueCount
        edges {
          cursor
          pullRequest: node {
            ... on PullRequest {
              id
              number
              title
              url
              createdAt
              author {
                login
                url
              }
              repository {
                owner {
                  login
                  url
                }
                name
                url
              }
            }
          }
        }
      }
    }
  `
  return graphqlRequest({
    query,
    variables: {
      githubQuery: join('state:open', githubQuery),
      endCursor,
      batchSize,
    },
    githubToken,
  })
}
