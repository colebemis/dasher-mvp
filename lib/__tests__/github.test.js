import { getGithubAuthUrl } from '../github'

jest.mock('next/config')

describe('getGithubAuthUrl', () => {
  it('returns correct URL', () => {
    expect(getGithubAuthUrl()).toMatchSnapshot()
  })
})
