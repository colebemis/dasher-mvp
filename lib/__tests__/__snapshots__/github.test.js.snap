// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getGithubAuthUrl returns correct URL 1`] = `"https://github.com/login/oauth/authorize?client_id=TEST_GITHUB_CLIENT_ID&redirect_uri=TEST_REDIRECT_URI"`;
